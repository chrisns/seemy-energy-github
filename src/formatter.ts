import { GetResponseDataTypeFromEndpointMethod } from '@octokit/types'
import { Octokit } from '@octokit/rest'
const octokit = new Octokit()

type PullRequestGetResponseDataType = GetResponseDataTypeFromEndpointMethod<typeof octokit.pulls.get>
type IssuesGetResponseDataType = GetResponseDataTypeFromEndpointMethod<typeof octokit.issues.get>

interface RecordPullRequest {
  repo: string
  id: number
  url: string
  holder: string
  author: string
  created_at: number
  closed_at: number
  merged_at: number
  assignees: number
  reviewers: number
  body_length: number
  time_to_merge: number
  merged_by: string
  commits: number
  review_comments: number
  additions: number
  deletions: number
  changed_files: number
}

export function formatPullRequest (pull: PullRequestGetResponseDataType): RecordPullRequest {
  return {
    repo: pull.base.repo.name,
    id: pull.number,
    url: pull.base.repo.url,
    holder: pull.base.user.login,
    author: pull.user && pull.user.login ? pull.user.login : 'unknown',
    commits: pull.commits,
    review_comments: pull.review_comments,
    additions: pull.additions,
    deletions: pull.deletions,
    changed_files: pull.changed_files,
    merged_by: pull.merged_by ? pull.merged_by.login : '',
    created_at: Date.parse(pull.created_at),
    closed_at: pull.closed_at ? Date.parse(pull.closed_at) : 0,
    merged_at: pull.merged_at ? Date.parse(pull.merged_at) : 0,
    assignees: pull.assignees ? pull.assignees.length : 0,
    reviewers: pull.requested_reviewers ? pull.requested_reviewers.length : 0,
    body_length: pull.body ? pull.body.length : 0,
    time_to_merge: pull.merged_at ? Date.parse(pull.merged_at) - Date.parse(pull.created_at) : 0,
  }
}

interface RecordIssue {
  repo: string
  id: number
  url: string
  holder: string
  author: string
  created_at: number
  closed_at: number
  assignees: number
  body_length: number
  time_to_close: number
  closed_by: string
  comments: number
}

export function formatIssue (issue: IssuesGetResponseDataType): RecordIssue {
  return {
    repo: issue.repository_url.split('/').reverse()[0],
    id: issue.number,
    url: issue.repository_url,
    holder: issue.repository_url.split('/').reverse()[1],
    author: issue.user && issue.user.login ? issue.user.login : 'unknown',
    comments: issue.comments,
    closed_by: issue.closed_by ? issue.closed_by.login : '',
    created_at: Date.parse(issue.created_at),
    closed_at: issue.closed_at ? Date.parse(issue.closed_at) : 0,
    assignees: issue.assignees ? issue.assignees.length : 0,
    body_length: issue.body ? issue.body.length : 0,
    time_to_close: issue.closed_at ? Date.parse(issue.closed_at) - Date.parse(issue.created_at) : 0,
  }
}
