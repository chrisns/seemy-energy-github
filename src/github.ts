import { Octokit } from '@octokit/rest'
import { createAppAuth } from '@octokit/auth-app'
import { formatPullRequest, formatIssue } from './formatter'
const { paginateRest } = require('@octokit/plugin-paginate-rest')
import { Endpoints } from '@octokit/types'

import SQS from 'aws-sdk/clients/sqs'

import DynamoDB from 'aws-sdk/clients/dynamodb'

export function getAuthenticatedOctokit (installationId: number): Octokit {
  return new Octokit({
    authStrategy: createAppAuth,
    auth: {
      appId: process.env.appId,
      type: 'app',
      privateKey: process.env.privateKey,
      installationId: installationId,
    },
  }).plugin(paginateRest)
}

interface sqsPullMessage {
  owner: string
  repo: string
  pull_number: number
  installation_id: number
}

export async function queryRepo (owner: string, repo: string, octokit: Octokit) {
  for await (const pulls of octokit.paginate.iterator(octokit.pulls.list, {
    owner: owner,
    repo: repo,
  })) {
    return await pulls.data.map(pullListPRtoSQS)
  }
}

export async function pullListPRtoSQS (
  pull: Endpoints['GET /repos/{owner}/{repo}/pulls']['response']['data'][0],
): Promise<SQS.SendMessageResult> {
  return sqsClient
    .sendMessage({
      MessageBody: JSON.stringify(<sqsPullMessage>{
        owner: pull.base.user.login,
        pull_number: pull.number,
        repo: pull.base.repo.url,
        installation_id: 1234,
      }),
      MessageDeduplicationId: `${pull.base.user.login}/${pull.base.repo.url}/${pull.number}`,
      MessageGroupId: `${pull.base.user.login}/${pull.base.repo.url}`,
      QueueUrl: process.env.PR_QUEUE,
    })
    .promise()
}

const config = {
  convertEmptyValues: true,
  ...(process.env.MOCK_DYNAMODB_ENDPOINT && {
    endpoint: process.env.MOCK_DYNAMODB_ENDPOINT,
    sslEnabled: false,
    region: 'local',
  }),
}

export const documentClient = new DynamoDB.DocumentClient(config)

export const sqsClient = new SQS()

const upsert_table = async (payload, table) =>
  documentClient
    .put({
      Item: payload,
      TableName: table,
    })
    .promise()

export async function ETLPullRequest (
  owner: string,
  repo: string,
  pull_number: number,
  octokit: Octokit,
): Promise<void> {
  const pull = await octokit.pull.get({ owner, repo, pull_number })
  const formatted = formatPullRequest(pull)
  await upsert_table(formatted, 'pull')
  return
}

export async function ETLIssue (owner: string, repo: string, issue_number: number, octokit: Octokit): Promise<void> {
  const issue = await octokit.issue.get({ owner, repo, issue_number })
  const formatted = formatIssue(issue)
  await upsert_table(formatted, 'issue')
  return
}
