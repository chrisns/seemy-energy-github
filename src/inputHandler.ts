import * as github from './github'
import { SQSEvent, APIGatewayProxyEventV2, APIGatewayProxyResultV2 } from 'aws-lambda'

export async function sqsPRQueueHandler (event: SQSEvent): Promise<void> {
  await event.Records.forEach(async record => {
    const body = <github.sqsPullMessage>JSON.parse(record.body)
    const octokit = github.getAuthenticatedOctokit(body.installation_id)
    await github.ETLPullRequest(body.owner, body.repo, body.pull_number, octokit)
  })
  return
}

export async function sqsIssueQueueHandler (event: SQSEvent): Promise<void> {
  await event.Records.forEach(async record => {
    const body = <github.sqsIssueMessage>JSON.parse(record.body)
    const octokit = github.getAuthenticatedOctokit(body.installation_id)
    await github.ETLPullRequest(body.owner, body.repo, body.issue_number, octokit)
  })
  return
}

export async function httpQueryRepoHandler (event: APIGatewayProxyEventV2): Promise<APIGatewayProxyResultV2> {
  const octokit = github.getAuthenticatedOctokit(Number(event.pathParameters.installationId))
  await github.queryRepo(event.pathParameters.owner, event.pathParameters.repo, octokit)
  return
}
