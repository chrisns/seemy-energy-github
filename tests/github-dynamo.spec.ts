jest.mock('@octokit/rest', () => {
  return {
    Octokit: jest.fn().mockImplementation(() => {
      return {
        issues: {
          get: () => fixtures.issueGood,
        },
        pulls: {
          get: () => fixtures.pullGood,
        },
      }
    }),
  }
})

import { setup } from 'jest-dynalite'
setup(__dirname)
import 'jest-dynalite/withDb'
import { Octokit } from '@octokit/rest'

import fixtures from './fixtures'

import * as mod from '../src/github'

const octokit = new Octokit()

describe('github dynamo', () => {
  test('ETLPullRequest queries pull request, persists to dynamodb', async () => {
    await expect(mod.ETLPullRequest('kubernetes', 'k8s.io', 1513, octokit)).resolves.toBeUndefined()
    const get = await mod.documentClient
      .get({
        TableName: 'pull',
        Key: {
          url: fixtures.pullGood.base.repo.url,
          id: fixtures.pullGood.number,
        },
      })
      .promise()
    return expect(get).toMatchSnapshot()
  })

  test('ETLIssue queries issue, persists to dynamodb', async () => {
    await expect(mod.ETLIssue('kubernetes', 'kubernetes', 56903, octokit)).resolves.toBeUndefined()
    const get = await mod.documentClient
      .get({
        TableName: 'issue',
        Key: {
          url: fixtures.issueGood.repository_url,
          id: fixtures.issueGood.number,
        },
      })
      .promise()
    return expect(get).toMatchSnapshot()
  })
})
