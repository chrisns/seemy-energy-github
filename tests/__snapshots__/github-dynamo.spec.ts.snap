// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github dynamo ETLIssue queries issue, persists to dynamodb 1`] = `Object {}`;

exports[`github dynamo ETLIssue queries issue, persists to dynamodb 2`] = `
Object {
  "Item": Object {
    "assignees": 0,
    "author": "mikksoone",
    "body_length": 2733,
    "closed_at": 1550815042000,
    "closed_by": "thockin",
    "comments": 253,
    "created_at": 1512598487000,
    "holder": "kubernetes",
    "id": 56903,
    "repo": "kubernetes",
    "time_to_close": 38216555000,
    "url": "https://api.github.com/repos/kubernetes/kubernetes",
  },
}
`;

exports[`github dynamo ETLPullRequest queries pull request, persists to dynamodb 1`] = `Object {}`;

exports[`github dynamo ETLPullRequest queries pull request, persists to dynamodb 2`] = `
Object {
  "Item": Object {
    "additions": 10,
    "assignees": 9,
    "author": "palnabarun",
    "body_length": 336,
    "changed_files": 2,
    "closed_at": 1610132191000,
    "commits": 1,
    "created_at": 1610130783000,
    "deletions": 5,
    "holder": "kubernetes",
    "id": 1513,
    "merged_at": 1610132190000,
    "merged_by": "k8s-ci-robot",
    "repo": "k8s.io",
    "review_comments": 0,
    "reviewers": 2,
    "time_to_merge": 1407000,
    "url": "https://api.github.com/repos/kubernetes/k8s.io",
  },
}
`;
